### PR TITLE
Falcon40b attn mask optimization

### PR DESCRIPTION
**Pull request by djordje-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/9906

### Problem description
Attention mask is pushed to device every time forward pass is executed. 

### What's changed
Instead of pushing attn_mask every time in forward pass, push attn_mask in init phase and in forward pass clone mask and slice to seq_len of input prompt. 

### Checklist

- [x] frequent tests: https://github.com/tenstorrent/tt-metal/actions/runs/9906510178

- [x] unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/9906522668

- [x] demo tests: https://github.com/tenstorrent/tt-metal/actions/runs/9909865131

- [x] model perf: https://github.com/tenstorrent/tt-metal/actions/runs/9906516963

### CODE REVIEW

File: models/demos/t3000/falcon40b/tt/falcon_model.py [hunk 1]
```diff
@@ -157,31 +186,8 @@ def model_preprocessing(self, llm_mode, input_ids, kv_cache_len, num_input_token
...
 
-            attention_mask_bool = torch.ones(batch_size, 1, sequence_size, sequence_size, dtype=bool)
-            attention_mask_bool = attention_mask_bool.triu(diagonal=1)
-
-            attention_mask_memconfig = self.model_config["ATTN_MASK_MEMCFG"]
-            if attention_mask_memconfig.is_sharded():
-                attn_mask_shard_shape = attention_mask_memconfig.shard_spec.shape
-                attn_mask_shard_shape[-1] = sequence_size
-                attention_mask_memconfig.shard_spec.shape = attn_mask_shard_shape
+            tt_attention_mask = self.slice_attn_mask(sequence_size)
```
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_model.py:189]: Can we extend this for decode too?

File: models/demos/t3000/falcon40b/tt/falcon_model.py
Comment (johanna-rock-tt) [inline models/demos/t3000/falcon40b/tt/falcon_model.py:1]: Maybe also rename preprocessing to init_preprocessing here? Just to be consistent?

**Review by johanna-rock-tt — APPROVED** [review]
Except for comments below, looks good to me!

File: models/demos/t3000/falcon40b/tt/falcon_model.py (hunk 1 above)
Comment (djordje-tt) [reply models/demos/t3000/falcon40b/tt/falcon_model.py:189]: Not the same as for prefill so would leave it for another PR since do not have enough time at the moment since I've already run all the tests.
